### ISSUE TITLE
Auto Upload detecting deleted pictures

### ISSUE THREAD

**Issue by earliestearl** [nextcloud/android#5225]
### Is your feature request related to a problem? Please describe.
I was using Google Photos for the last few years. Now, coming to Nextcloud, I miss the chance to sort photos out after they have been taken. Photos are immediately uploaded but usually I review them a few hours later. Only the best photos should remain in the cloud. Bad photos locally deleted should also be deleted online.

### Describe the solution you'd like
Photos taken are automatically upload but when I delete some of them a few hours later, they are deleted in the cloud as well.

### Describe alternatives you've considered
At the moment, I am not aware of any workaround. Any ideas here?

### Additional context
Keep in mind that it still has to be possible to delete old photos locally from the phone to free up space. In this case, of course, all photos have to stay in the cloud.

Google solves that by providing an app for managing photos on the phone. Pictures can be view and deleted in this app. Deleted pictures are also deleted on Google Photos online. In addition, there is a function called "Free phone space" which suggests to delete all photos already uploaded.

**Comment by AndyScherzinger** [comment]
cc @tobiasKaminsky @nextcloud/designers @jancborchardt for feedback

**Comment by cieska** [comment]
me too, regarding googles photos functionality.

**Comment by earliestearl** [comment]
Meanwhile I found another solution for syncing photos. People with similiar needs should checkout [SyncThing](https://syncthing.net/). Peer-to-Peer synchronization. Totally different approach but it fits my personal needs and I stopped using NextCloud for photos.

However, NextCloud still is a great project and great platform. So I would definitely appreciate to see this feature someday.

**Comment by bclnr** [comment]
I use the photo auto-upload to have a local backup of my iCloud camera roll.
It should be possible to automatically delete photos that were deleted from the Photos app, to always keep Nextcloud in sync.

**Comment by ma12-co** [comment]
I think it would be a very nice feature to have as an opt in option, what do you think @jancborchardt?

**Comment by jancborchardt** [comment]
Totally agree, yep. This is one big reason why I actually do not use Automatic upload. This of course creates another issue because then you need to manually upload every once in a while, which is not nice either.

So it would definitely be nice, but is it doable or technically feasible @AndyScherzinger @tobiasKaminsky?

**Comment by AndyScherzinger** [comment]
> So it would definitely be nice, but is it doable or technically feasible @AndyScherzinger @tobiasKaminsky?

TL;DR: There would be ways to implement this but it would be quite the effort

----

There are two scenarios mentioned above: "manual deletions to be synced to server" (read: local deletion=>server-side deletion), "free up some space" (read: make sure files are available server-side then delete them locally).

Issues I think of:
* a manual mass deletion would lead to a mass deletion on the server so people already using the app likely won't expect this (and will loose their data; I am aware of the trashbin...) -> this would be handled via the "opt-in" character of the feature of course (still one might forget about it, like "usb connection, delete all photos...")
* We don't have a Nextcloud Photos app as-in any actions on images are handled through our app (if you delete stuff outside Google's photo app, they don't care, neither should we).

So in order to implement this we would need to:
* Extend the Auto Upload view(s) to browse all local images and visually show if available on server
* Add deletion action within such a view (might have issues with file permission, so this might not work or break based on (future) Android version(s)
* have a sync mechanism to even know which files have been uploaded already (we don't store this info yet and also don't have a check/compare mechanism as of today)

**Comment by ma12-co** [comment]
Sounds good, looking forward to see all this awesome photo goodness moved to a nextcloud photos android app :)

**Comment by tobiasKaminsky** [comment]
I think this is totally out of scope of the app.

Let us transfer this idea to Desktop:
- you have your image colletion in /home/user/tobi/images
- you add two images there
- Desktop client should detect this, upload it and then remove it, if you delete it there?
- in fact Desktop client is doing a total "stupid" 2 way sync of *any* file within its folder, but does not tamper at all with any file *outside* it

So to transfer it back, I would suggest this
- upload images via auto upload & move file inside NC folder
- have a real 2 way sync (we do not have right now)
- delete any file you wish -> gets also deleted on server

**Comment by AndyScherzinger** [comment]
> I think this is totally out of scope of the app.

I think it depends on the solution. We can't sync deletions because this is impossible in my opinion since the app can never "know" why a user deleted an image (like to free space on the phone).
So the way to properly detect a deletion (like you stated @tobiasKaminsky) is when such a deletion is triggered within the Nc app since we could then just ask the user what to do (delete local, remote, both; this is also how the Amazon app behaves). This would either have to be done like mentioned by @tobiasKaminsky _or_ we would have to implement browsing within the Auto Upload screen for each of these folders, implement multi-selection and a delete action that asks which version to be deleted (local/server/both). This would mean we need to store (for performance reasons, sync-check just takes too long) which file we already uploaded and indicate this (like done in Google Photos which indicated files that haven't been uploaded yet).
So the later solution is quite the effort I am afraid so unlikely to happen any time soon except if someone would be willing to implement this (I'd like to but don't have the time)

**Comment by ben-utzer** [comment]
> We can't sync deletions because this is impossible in my opinion since the app can never "know" why a user deleted an image (like to free space on the phone).

Why not just offer a two way sync? I don't get why this is taken so complicated, there is many cases where two way sync is useful and Nextcloud Android app is the first cloud storage app I ran into that does not offer this function. 
I read rumors that there was a tow way sync in the early times, I guess that is a false info. 

What we need is the following functions autoupload (upload only or send only), two way sync and download (again only download from the server).
That is precisely what a users would expect, what existing functions can be migrated to and what other cloud storage services offer since the 2010 or so.

**Comment by wis775** [comment]
As a user, I would expect a two way sync of auto uploaded files.

I‘ve set up the auto-upload feature, to have easy access to the photos of my mobile from my pc.

So, lets tell a short story.

On a trip, I take some photos with my mobile. After returning home, the auto-upload feature uploads the photos into the cloud and moves them into the app folder. On the next day, I have a look at the photos on my moblie (lets say with Samsung Gallery) and delete the „not-to-keep“ ones.

I would expect, that the deleted photos are also deleted in the cloud. Why I should delete my photos twice?

Why I use Samsung Gallery App and not the Nextcloud Client to show/delete photos: Because it’s integrated into the operating system, it can cast the photos to Chromecast and have the better usability.

**Comment by ben-utzer** [comment]
> As a user, I would expect a two way sync of auto uploaded files.
I can understand that and agree fully to what you wrote. 
Yesterday someone pointed out that I misunderstand the nextcloud client for Android, it is just a app you can use instead of the web interface, it is not a sync, download or upload client. So I wonder if that is right, if so there should be a separate sync client.

**Comment by weiss** [comment]
> So, lets tell a short story.

I think the problem is pretty much clear.  As has been pointed out above, the solution isn't totally obvious; and will either way require some development effort.

FWIW, my personal workaround for your use case is to work with two folders: The `DCIM/Camera` folder is _not_ auto-uploaded to Nextcloud.  I take photos into that folder and check them later on.  After deleting the “not-to-keep” pics, I move the remaining photos into a `Pictures/Album` folder which _is_ then auto-uploaded to Nextcloud.  (I.e., the `DCIM/Camera` folder is like a to-do list—it's empty once I'm done going through my recent photos.)

**Comment by ben-utzer** [comment]
This feature request should be changed to be more general, the requested feature should be considered for any type of files/folders. This is actually a request for more versatile synchronization between online folder and a local copy on the device. 

Again summarized possible implementation is to have 3 types of sync/upload/download: 
1. Only upload
2. Only Download
3. Both way sync

For 1 it would mean the master copy is on the mobile device, all changes from there will be uploaded, including deletions and changes. Online changes would be lost. 
For 2 it would mean the master copy is online on the server. Any change on the server would be downloaded to the mobile device. Any change on the mobile device would be lost. 
For 3 it would mean that any change online or on the mobile device would sync. This requires some kind of conflict management, if client recognizes local and online change on a file it has to create a conflict copy and the user must deal with it, older change file could be renamed $filename_conflict_$date_$time.$suffix
Users can select the behaviour of the synchronization.

**Comment by unpairedliabilitylibrarian** [comment]
Hi guys,
I'm giving nextcloud (again) a new try and I'm currently using syncthing because it has this two way sync with a optional "delete on the other site if deleted on this site" feature. I'd really like to see this in the nextcloud app - otherwise a auto-upload of my pictures taken on my phone would result in double work: delete pictures on the mobile phone and then delete on the server manually.

Thank you :)

**Comment by zilexa** [comment]
> Hi guys,
> I'm giving nextcloud (again) a new try and I'm currently using syncthing because it has this two way sync with a optional "delete on the other site if deleted on this site" feature. I'd really like to see this in the nextcloud app - otherwise a auto-upload of my pictures taken on my phone would result in double work: delete pictures on the mobile phone and then delete on the server manually.
> 
> Thank you :)
> 
> ![Screenshot_20210327-075946](https://user-images.githubusercontent.com/43911405/112713010-19619000-8ed3-11eb-8f3e-02b28f36f112.png)

I highly recommend you to try FolderSync instead. Syncthing works fine for pure & 100% 2-way sync. If you need to disable syncing of delete actions or if you want 1-way sync with/without delete actions, you will hit a wall: the Syncthing UI will state out of sync errors and (if you ever disable syncing of deletes for a single folder) will never show finished sync actions (always stuck at <100%).

I have discussed this thoroughly on their forum with their Devs and this is how they like it: the UI provides zero useful information when a user messes with the 2 mentioned options.

FolderSync app allows clear sync for 1-way and 2-way scenarios with or without delete actions, both at scheduled times and realtime. 

If NextCloud would implement that, I would instantly use the NextCloud app. By default I only use CX File Explorer (it has fantastic user friendly interface, webDAV support and is free, also ad-free) for browsing files and FolderSync for syncing.

**Comment by loligans** [comment]
I like the idea of a NextCloud photos app. I think it could be a good solution to a few problems. Changing Auto Sync behavior in the existing NextCloud app will likely ruin someones day. We should try to avoid that as much as possible.

A NextCloud Photos app could:
1. Get the consent of a user before deleting image(s).
2. Permit (or restrict) image alterations and sync the changes automatically.
3. Include NextCloud specific sharing capabilities.

**Comment by mrtnd** [comment]
if there was i way to make the photos / media tab to behave like google photos / ios photos on deletion would be great. That would save a lot of redundant work.

**Comment by cieska** [comment]
> if there was i way to make the photos / media tab to behave like google photos / ios photos on deletion would be great. That would save a lot of redundant work.

Like how?

**Comment by mrtnd** [comment]
> > if there was i way to make the photos / media tab to behave like google photos / ios photos on deletion would be great. That would save a lot of redundant work.
> 
> Like how?

when one deletes photos using google photos or other Gallery app those changes are not synced to the server. Deleted photos are still presented on the server and you need to delete them manually.

**Comment by loligans** [comment]
> if there was i way to make the photos / media tab to behave like google photos / ios photos on deletion would be great. That would save a lot of redundant work.

I think NextCloud should treat all deletions from non-NextCloud apps as "local delete" operations only. If someone wants to permanently delete an image from the server they would need to go through a NextCloud app and explicitly choose delete permanently (from the server).

**Comment by derykl** [comment]
Yes, in light of recent revelations of even Apple throwing away the notion of privacy, I'm trying to move to private and independent solutions rather than relying on Apple or Google and replacing the photo apps is a major pain point right now.  I'm used to the behaviour on Android that you'd use Google Photos as the gallery app and deleting from that deletes from local, or the behaviour on iOS where it simply deletes on both with a warning.

FolderSync is so close to being a reasonable compromise, except it doesn't seem to work with e2e encrypted folders.  As I'm using a 'cloud' server that I don't own, I'd rather they stay encrypted, so it's almost but not quite the right solution...

**Comment by GambaJo** [comment]
two way sync would be nice. especially for the auto upload folder.

**Comment by bes1002t** [comment]
> Hi guys,
> I'm giving nextcloud (again) a new try and I'm currently using syncthing because it has this two way sync with a optional "delete on the other site if deleted on this site" feature. I'd really like to see this in the nextcloud app - otherwise a auto-upload of my pictures taken on my phone would result in double work: delete pictures on the mobile phone and then delete on the server manually.
> 
> Thank you :)
> 
> ![Screenshot_20210327-075946](https://user-images.githubusercontent.com/43911405/112713010-19619000-8ed3-11eb-8f3e-02b28f36f112.png)

I would recommend to open a new ticket for a new request. If I understand correctly your initial request was a "sort by date taken" feature for photos. I would love such an feature as well, but we should keep the tickets clean, otherwhise nobody will read it.

**Comment by utzer** [comment]
> The other way if all deletes on the phone are synced to the nextcloud will be a critical issue when phone memory fills up, then it will be even more complicated.

Therefor there should be Folder that are upload only folder, which is the case right now, and sync folders, which sync all changes between every sync peer.

**Comment by icasdri** [comment]
I agree that two-way sync would be nice to have as an option. However, when this is implemented, can this please be labelled clearly as "Two-Way Sync" separate form "Auto Upload"? To me "Auto Upload" implies one-way upload (no synced deletions), and as some have mentioned above, I do deletions manually on device when device storage fills up and I do *not* expect them to be deleted from the server. If the default behavior or the behavior labelled as "Auto Upload" changes, users *will* lose data.

**Comment by stephen304** [comment]
I would love for a feature similar to the google photos "free up space". I like the auto sync because I often grab the photo from my computer when it syncs, and I only need photos on the phone for a short time after taking them if I want to send or post it immediately.

There is an option to delete photos after uploading, which I may use, but it will be slightly annoying if I want to share something I just took since I will have to download it again. I prefer just manually running a cleanup task at my own discretion.

**Comment by bonswouar** [comment]
Adding that feature as an option would be really great!
And as @utzer pointed out earlier it should be more general, for any sync folder, not just Photos. (could we change the title of the issue maybe?)

One useful case would be to automatically remove old backups generated with https://github.com/seedvault-app/seedvault . Right now they're deleted locally but stays on NC until you delete them manually (although it seems the vault directory isn't listed in Nextcloud's app auto-sync folders, so I guess it should also be implemented ino)

**Comment by PargsZ** [comment]
Any news on this? This feature would be really great!
